src/lib/group_map/group_map.ml

```diff
@@ -24,323 +30,69 @@ struct
-          ; ((0, 10), "1") ] }
-
-  let%test_unit "sanity check" =
-    let {n1; d1; n2; d2; n3; d31; d32} = t in
-    List.iter [n1; d1; n2; d2; n3; d31; d32] ~f:(fun t ->
-        Hashtbl.iter_keys t ~f:(fun (a, b) ->
-            let three_j = (2 * a) + (3 * b) in
-            if three_j mod 3 <> 0 then
-              failwithf "2 * %d + 3 * %d = %d = %d mod 3" a b three_j
-                (three_j mod 3) ()
-            else () ) )
-
-  module Magic_numbers = struct
-    type nonrec 'f t = 'f AB_pair.Table.t coefficients
-
-    let create ~negate ~of_string : _ t =
-      map t
-        ~f:
-          (Hashtbl.map ~f:(fun s ->
-               if s.[0] = '-' then
-                 let s' = String.sub s ~pos:1 ~len:(String.length s - 1) in
-                 negate (of_string s')
-               else of_string s ))
-  end
-
-  let ( @! ) tbl (a, b) =
-    match Hashtbl.find tbl (a, b) with
-    | None ->
-        failwithf "group_map: coefficient for (%d, %d) not found" a b ()
-    | Some c ->
-        c
-
-  let abs j =
-    (* Find all integers a, b >= 0 such that 2 a + 3 b = 3 j.
-        Note that this implies 
-        3 b <= 3 j so 0 <= b <= j.
-    *)
-    let open Sequence in
-    let open Let_syntax in
-    let%bind b = interval 0 j in
-    let two_a = (3 * j) - (3 * b) in
-    if two_a mod 2 = 0 then return (two_a / 2, b) else empty
-
-  module Intermediate = struct
-    type 'f t = {ab_products: 'f AB_pair.Table.t}
-
-    let max_j = 16
-
-    let max_a = 3 * max_j / 2
-
-    let max_b = max_j
-
-    let create (type t) (module F : Field_intf.S with type t = t) ~a:coeff_a
-        ~b:coeff_b =
-      let mul = F.( * ) in
-      let one = F.one in
-      let a_powers = powers ~mul ~one coeff_a max_a in
-      let b_powers = powers ~mul ~one coeff_b max_b in
-      let ab_products =
-        let res = AB_pair.Table.create () in
-        let open Sequence in
-        iter
-          (concat_map (interval 0 max_j) ~f:abs)
-          ~f:(fun (a, b) ->
-            Hashtbl.find_or_add res (a, b) ~default:(fun () ->
-                F.(a_powers.(a) * b_powers.(b)) )
-            |> ignore ) ;
-        res
-      in
-      {ab_products}
-  end
-
-  let init_range x y ~f = Array.init (y - x + 1) ~f:(fun i -> f (x + i))
-
-  type 'f t = {coefficients: 'f array coefficients; a: 'f; b: 'f; a2: 'f}
+  type 'f t =
+    { u: 'f
+    ; u_over_2: 'f
+    ; projection_point: 'f Conic.t
+    ; conic_c: 'f
+    ; a: 'f
+    ; b: 'f }
   [@@deriving fields]
 
-  (* Compute the coefficients for all the magic polynomials *)
-  let create (type t) ((module F : Field_intf.S with type t = t) as m) ~a ~b =
+  let create (type t) (module F : Field_intf.S_unchecked with type t = t) ~a ~b
+      =
     let open F in
-    let {Intermediate.ab_products} = Intermediate.create m ~a ~b in
-    let {n1; d1; n2; d2; n3; d31; d32} =
-      Magic_numbers.create ~negate
-        ~of_string:(Fn.flip Sexp.of_string_conv_exn t_of_sexp)
+    let first_map f =
```

Comment:
maybe note that `first` here means choosing the first (counting upwards from 0) `u` that satisfies the constraints for this subproblem, so as to separate it mentally from choosing the first of x1, x2, x3? 

also does we have an expected running time for this? i don't think it'll ever be a problem but would be curious. also why does this `u` not depend eg on the output of a hash function?

I'm also a little confused where/how the output will change on different inputs if finding `u` isn't a function of some input parameter?

Reply:
This `u` is one of the parameters for the map `field -> V`. 

I.e., there is a family of maps `field -> V`, one for each valid value of type `field Params.t`, which is the record
```ocaml
    { u: field
    ; u_over_2: field
    ; projection_point: field Conic.t
    ; conic_c: field
    ; a: field
    ; b: field }
```

Such a value is "valid" if it satisfies all the conditions in the paper. Namely: `u` satisfies all those constraints, `conic_c` is the coefficient of `y^2` in the conic they give, and `projection_point` is a point on that conic.

Reply:
Also these parameters are computed at compile time and `u = 1` works so speed is not an issue.

Reply:
I.e., this code runs and picks `u = 1` every time.

Reply:
thanks for the explanation and for the new comments!